returnn/tf/layers/rec.py

```diff
@@ -3294,10 +3303,13 @@ def get_sub_layer(self, layer_name):
     assert res, "Could not get out data for sub-layer template {}.".format(full_layer_name)
     output, network, sub_layer_class = res
 
-    sub_layer_template = _TemplateLayer(self.network, full_layer_name)
+    # During template construction templates might be re-initialized by calling init() several times. For sub-layers
+    # we have to do the same, so we have to store the sub-layers created so far.
```

Comment:
Why do the sub layer templates need to be stored?

Reply:
Because of this trial and error approach of `_construct_template()`, `get_sub_layer()` might be called several times.
Then in `move_outside_loop` there is the check:
```
assert self.layer_data_templates[layer.name] is layer
```
where `layer` is taken from the layer `_TemplateLayer.dependencies` of another layer. If we create a new `_TemplateLayer` in each `get_sub_layer` call this check would fail, because we have created a new object. We have to update the existing object instead, also because it is referenced in other layers in `dependencies`.

Reply:
I see, the `... is layer` check. Is this the only place where we expect the same instance?

Anyway, this is relevant for the comment. This is the crucial information you need to explain in the comment, why we cannot recreate the instances. (And just saying "because we have some `assert` for it elsewhere" does not explain anything. Then the question is, why do we have the `assert`...)

Reply:
> Is this the only place where we expect the same instance?

It seems to be the only explicit check, yes. Moving out layers could be done by checking dependencies via the layer name only, so without the actual instance (although sometimes we check for the name, sometimes for the instance in the current code). But I think to get search choices we need the dependency instances, and probably if it's not the latest correct ones this will cause bugs. So I think the check is good.

But yes, let me update the comment.